### PR TITLE
Add `@throttle` directive to set field rate limit using Laravel rate limiting services

### PR DESCRIPTION
**Pull request by k0ka**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

Adds `@throttle` directive. It is adapted from Laravel and works on per-field basis.

### CODE REVIEW

File: src/Schema/Directives/ThrottleDirective.php [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    public function handleField(FieldValue $fieldValue, Closure $next): FieldValue
+    {
+        $originalResolver = $fieldValue->getResolver();
+
+        $limits = [];
+        $name = $this->directiveArgValue('name', null);
+        if ($name !== null) {
+            if (! method_exists($this->limiter, 'limiter')) {
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:79]: This validation can happen in `manipulateField()` and thus happen at schema build time.

File: src/Schema/Directives/ThrottleDirective.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+) on FIELD_DEFINITION
+GRAPHQL;
+    }
+
+    public function handleField(FieldValue $fieldValue, Closure $next): FieldValue
+    {
+        $originalResolver = $fieldValue->getResolver();
+
+        $limits = [];
+        $name = $this->directiveArgValue('name', null);
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:77]:
```suggestion
        $name = $this->directiveArgValue('name') ?? null;
```

File: src/Schema/Directives/ThrottleDirective.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                throw new DirectiveException('Named limiter requires Laravel 8.x or later');
+            }
+
+            $limiter = $this->limiter->limiter($name);
+            /** @phpstan-ignore-next-line $limiter may be null although it's not specified in limiter() PHPDoc */
+            if (is_null($limiter)) {
+                throw new DirectiveException("Named limiter $name is not found.");
+            }
+
+            $limiterResponse = call_user_func($limiter, $this->request);
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:89]:
```suggestion
            $limiterResponse = $limiter($this->request);
```

File: src/Schema/Directives/ThrottleDirective.php [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            if (class_exists(Unlimited::class) && $limiterResponse instanceof Unlimited) {
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:90]: Won't this fail if the class does not exist? Maybe we have to use a string instead.

File: src/Schema/Directives/ThrottleDirective.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                return $next($fieldValue);
+            }
+
+            if ($limiterResponse instanceof Response) {
+                throw new DirectiveException(
+                    "Named limiter $name returns instance of Responst which is not supported."
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:96]:
```suggestion
                    "Expected named limiter {$name} to return an array, got instance of " . get_class($limiterResponse)
```

**Comment by k0ka** [comment]
I had to add a lot of `@phpstan-ignore-next-line` in tests, because phpstan doesn't understand that code won't be executed for old versions of Laravel.

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        );
+    }
+
+    public function testNamedLimiter(): void
+    {
+        /** @var RateLimiter $rateLimiter */
+        $rateLimiter = $this->app->make(RateLimiter::class);
+
+        if (! method_exists($rateLimiter, 'for') || ! class_exists('Illuminate\Cache\RateLimiting\Limit')) {
+            return;
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:70]: You can use `$this->markTestSkipped()` and give a reason. Look around for other uses to fix the resulting PHPStan error.

**Comment by spawnia** [comment]
> I had to add a lot of `@phpstan-ignore-next-line` in tests, because phpstan doesn't understand that code won't be executed for old versions of Laravel.

That is fine. Annoying, but worth the hassle to prevent serious bugs. Thanks for digging into this.

File: src/Schema/Directives/ThrottleDirective.php (hunk 4 above)
Comment (k0ka) [reply src/Schema/Directives/ThrottleDirective.php]: I've removed this check - it is not needed in production environment. This class always exists in Laravel 8

File: src/Schema/Directives/ThrottleDirective.php (hunk 1 above)
Comment (k0ka) [reply src/Schema/Directives/ThrottleDirective.php]: yes, moved this these checks to `manipulateField()`

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php (hunk 6 above)
Comment (k0ka) [reply tests/Integration/Schema/Directives/ThrottleDirectiveTest.php]: Yep, marked tests as skipped and also used `AppVersion::below(8.0)` to check version

File: docs/master/api-reference/directives.md [hunk 7]
```diff
@@ -2592,6 +2592,41 @@ type Subscription {
 }
 ```
 
+## @throttle
+
+```graphql
+"""
+Sets rate limit to access the field, just like Laravel's `ThrottleRequests` middleware.
+"""
+directive @throttle (
```
Comment (spawnia) [inline docs/master/api-reference/directives.md:2601]:
```suggestion
directive @throttle(
```

File: src/Schema/Directives/ThrottleDirective.php [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        $this->request = $request;
+    }
+
+    public static function definition(): string
+    {
+        return <<<'GRAPHQL'
+"""
+Sets rate limit to access the field. Does the same as ThrottleRequests Laravel Middleware.
+"""
+directive @throttle (
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:48]:
```suggestion
directive @throttle(
```

File: src/Schema/Directives/ThrottleDirective.php [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+                'decayMinutes' => $this->directiveArgValue('decayMinutes', 1),
+            ];
+        }
+
+        return $next(
+            $fieldValue->setResolver(
+                function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo) use (
+                    $originalResolver,
+                    $limits
+                ) {
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:115]:
```suggestion
                function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo) use ($originalResolver, $limits) {
```

File: src/Schema/Directives/ThrottleDirective.php [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+                        );
+                    }
+
+                    return $originalResolver($root, $args, $context, $resolveInfo);
+                }
+            )
+        );
+    }
+
+    public function manipulateFieldDefinition(DocumentAST &$documentAST, FieldDefinitionNode &$fieldDefinition, ObjectTypeDefinitionNode &$parentType)
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:130]:
```suggestion
    public function manipulateFieldDefinition(DocumentAST &$documentAST, FieldDefinitionNode &$fieldDefinition, ObjectTypeDefinitionNode &$parentType): void
```

File: src/Schema/Directives/ThrottleDirective.php [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        $name = $this->directiveArgValue('name') ?? null;
```
Comment (spawnia) [inline src/Schema/Directives/ThrottleDirective.php:132]:
```suggestion
        $name = $this->directiveArgValue('name');
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+use Nuwave\Lighthouse\Support\AppVersion;
+use Tests\TestCase;
+use Tests\Utils\Queries\Foo;
+
+class ThrottleDirectiveTest extends TestCase
+{
+    public function testWrongLimiterName(): void
+    {
+        $this->schema = /** @lang GraphQL */
+            '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:19]:
```suggestion
        $this->schema = /** @lang GraphQL */ '
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        type Query {
+            foo: Int @throttle(name: "test")
+        }
+        ';
+
+        $this->expectException(DefinitionException::class);
+        $this->graphQL(
+/** @lang GraphQL */ '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:27]:
```suggestion
        $this->graphQL(/** @lang GraphQL */ '
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            foo
+        }
+        '
+        );
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:32]:
```suggestion
        ');
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public function testNamedLimiterReturnsRequest(): void
+    {
+        if (AppVersion::below(8.0)) {
+            $this->markTestSkipped('Version less than 8.0 does not support named requests.');
+        }
+
+        $this->schema = /** @lang GraphQL */
+            '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:42]:
```suggestion
        $this->schema = /** @lang GraphQL */ '
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        $rateLimiter->for(
+            'test',
+            function () {
+                return response('Custom response...', 429);
+            }
+        );
+
+        $this->expectException(DirectiveException::class);
+        $this->graphQL(
+/** @lang GraphQL */ '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:60]:
```suggestion
        $this->graphQL(/** @lang GraphQL */ '
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+            foo: Int @throttle(name: "test")
+        }
+        ';
+
+        /** @var RateLimiter $rateLimiter */
+        $rateLimiter = $this->app->make(RateLimiter::class);
+        $this->assertTrue(method_exists($rateLimiter, 'for'));
+        $rateLimiter->for(
+            'test',
+            function () {
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:53]:
```suggestion
            static function () {
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        );
+
+        $this->expectException(DirectiveException::class);
+        $this->graphQL(
+/** @lang GraphQL */ '
+        {
+            foo
+        }
+        '
+        );
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:65]:
```suggestion
        ');
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public function testNamedLimiter(): void
+    {
+        if (AppVersion::below(8.0)) {
+            $this->markTestSkipped('Version less than 8.0 does not support named requests.');
+        }
+
+        $this->schema = /** @lang GraphQL */
+            '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:75]:
```suggestion
        $this->schema = /** @lang GraphQL */ '
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+            foo: Int @throttle(name: "test")
+        }
+        ';
+
+        /** @var RateLimiter $rateLimiter */
+        $rateLimiter = $this->app->make(RateLimiter::class);
+        $this->assertTrue(method_exists($rateLimiter, 'for'));
+        $rateLimiter->for(
+            'test',
+            function () {
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:86]:
```suggestion
            static function () {
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+                /** @phpstan-ignore-next-line phpstan ignores markTestSkipped */
+                return Limit::perMinute(1);
+            }
+        );
+
+        $this->graphQL(
+/** @lang GraphQL */ '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:93]:
```suggestion
        $this->graphQL(/** @lang GraphQL */ '
```

File: tests/Integration/Schema/Directives/ThrottleDirectiveTest.php [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            foo
+        }
+        '
+        )->assertJson(
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/ThrottleDirectiveTest.php:98]:
```suggestion
        ')->assertJson(
```

**Review by spawnia — APPROVED** [review]

**Pull request merged**